which.1
.TH WHICH 1
.rn RS Rs
.de RS
.na
.nf
.Rs
..
.rn RE Re
.de RE
.Re
.fi
.ad
..
.de Sx
.PP
.ne \\$1
.RS
..
.de Ex
.RE
.PP
..
.na
.SH NAME
which \- show full path of commands
.SH SYNOPSIS
.B which
[options] [\-\-] programname [...]
.SH DESCRIPTION
.BR Which
takes one or more arguments.  For each of its arguments it
prints to stdout the full path of the executables that would have
been executed when this argument had been entered at the shell prompt.
It does this by searching for an executable or script in the
directories listed in the environment variable
.B PATH
using the same algorithm as bash(1).
.SH OPTIONS
.TP 4
.B "\-\-version"
Print version information on standard output then exit successfully.
.TP
.B "\-\-skip-dot"
Skip directories in
.B PATH
that start with a dot.
.TP
.B "\-\-skip-tilde"
Skip directories in
.B PATH
that start with a tilde and executables which reside in the HOME directory.
.TP
.B "\-\-show-dot"
If a directory in
.B PATH
starts with a dot and a matching executable was found for that path,
then print "./programname" rather than the full path.
.TP
.B "\-\-show-tilde"
Output a tilde when a directory matches the
.B HOME
directory.  This option is ignored when
.B which
is invoked as root.
.TP
.B "\-\-tty\-only"
Stop processing options on the right if not on tty.
.TP
.B "\-\-all, \-a"
Print
.I all
matching executables in PATH, not just the first.
.SH RETURN VALUE
.BR Which
returns the number of failed arguments, or -1 when no
.I programname
was given.
.SH EXAMPLE
A useful way to use this utility is by adding an alias for
.B which
like the following:
.Sx 1
alias which='which --tty-only --show-tilde --show-dot'
.Ex
This will print the readable ~/ and ./ when starting which from your prompt,
while still printing the full path when used from a script:
.Sx 4
> which q2
~/bin/q2
> echo `which q2`
/home/carlo/bin/q2
.RE
.SH BUGS
If
.I programname
is an alias which starts a different executable, then
.B which
will give the wrong path; it doesn't do alias expansion.
The HOME directory is determined by looking for the HOME environment
variable,
.B which
aborts when this variable doesn't exist.
.B Which
will consider two equivalent directories to be different when one
of them contains a path with a symbolic link.
.SH AUTHOR
Carlo Wood <carlo@gnu.org>
.SH "SEE ALSO"
bash(1)


NEWS
Version 2.21 --> 2.40

* Windows support improved


Version 2.20 --> 2.21

* Upgraded code from bash to version 4.3 (now uses eaccess).
* Fixed a bug related to getgroups / sysconfig that caused Which not to see
  more than 64 groups for a single user (thanks to Dmitry Bakshaev for finding
  and analyzing this).
* Build system maintenance.

Version 2.19 --> 2.20

* xmalloc.h included a readline header for no apparent reason. The compilation
  failed if readline was not installed. The dependency on the readline header
  was removed. Submitted by Tushar Teredesai.

Version 2.18 --> 2.19

* Upgraded code from bash to version 3.2. This DOES has influence
  on how Which behaves under certain circumstances.
* Support for Cygwin: Two leading slashes are now treated as
  possibly special, because a double leading slash can have an
  implementation-defined meaning (depending on the Operating
  System), as is for example the case in Cygwin.
  See http://www.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap04.html#tag_04_11
  This means that GNU which no longer collapses two leading slashes
  of a path into a single slash.
* When the environment variable HOME is not set, the home directory
  is now read from /etc/passwd or set to '/' if no home directory
  could be found (this now matches the tilde lib used in bash).
* Fixed the function declaration given in README.alias
* Added a @dircategory and @direntry to which.texi
* Changed the license to GPLv3.

Version 2.17 --> 2.18

* Fixes regarding cwautomacros (maintainer build system only).
* Improvement of 'make release' target (maintainer target).

Version 2.16 --> 2.17

* Add support for shells that output '{' on the same line as the
  function name in the output of 'declare -f' (ie zsh).
* Added support for Tandem OSS.
* Build system now uses cwautomacros (mostly just it's autogen.sh).

Version 2.15 --> 2.16

* This release fixes the bug, introduced in 2.15, that files for
  which the 'other' permission bits apply do NOT need to be readable.
  (The confusion came from the fact that it needs to be readable
  when it is a shell script but there is no way to find out if a
  non-readable file is a shell script or not.)
  The next release of bash will be using the same algorithm again,
  and that is the only thing that really matters.  This is also the
  algorithm that most other shells use.

Version 2.14 --> 2.15

* Upgraded build system to work with automake 1.7.2 and autoconf 2.57.
* Fixed search routine to skip executables that are not executable.
  This is NOT done by all versions of bash (which is a bug).
  The stock bash-2.0.5b thinks it is allowed to execute a file
  when it has the eXec bit set for our group or other, but that
  is not the case: when we own the file, the eXec bit of the *owner*
  must be set and the group/other bits are irrelevant.  And when
  we don't own the file, but the file is in our group, then only
  the group bit matters and not the other eXec bit.  And finally,
  when we do not own the file and the file is not in our group
  then not only the 'other' eXec bit needs to be set, but the
  file must be readable as well.  This has been reported to the
  maintainer of bash.  For the rest, the search routine has been
  copied/upgraded to that of bash version 2.0.5b.

Version 2.13 --> 2.14

* Upgraded build system to work with automake 1.5 and autoconf 2.53.
* Upgraded --read-functions functionality to work with bash 2.0.5b.

Version 2.12 --> 2.13

* Support for shell functions.  This version also fixes the
  problem that the alias for `which' wasn't inherited by subshells.
  You need to change the alias by a shell function! (The rpm does
  this automatically).
* Fixes for the .spec file (merged changes from RedHat 7.1)
  Now installs the which.info file as well.

Version 2.11 --> 2.12

* Minor man page fix (no reason to upgrade at all thus)

Version 2.10 --> 2.11

* Added a texinfo file (no reason to upgrade at all thus)
  `info' is the prefered format for documenting GNU tools.

Version 2.9 --> 2.10

* Buildsystem fixes (no reason to upgrade at all thus):
  Used the correct order of calling the automake/conf utils,
  fixed my E-mail address and added a Source: line to the rpm spec.

Version 2.8 --> 2.9

* Fixed compile problem in Solaris 5.6 using cc
* Added a `--help' option
* RPMs now add a /etc/profile.d script, automatically setting a useful alias

Version 2.7 --> 2.8

* aclocal.m4 was missing from the tar, resulting in
  a build failure if autoconf isn't installed.

Version 2.6 --> 2.7

* Support for aliases
* Configure/compile fix in the `tilde' directory.

Version 2.5 --> 2.6

* Doesn't depend on libreadline.a anymore.

Version 2.4 --> 2.5

* Fixed which-2.spec
* RPMs are now available from http://www.xs4all.nl/~carlo17/which


which.c
/*
 * which v2.x -- print full path of executables
 * Copyright (C) 1999  Carlo Wood <carlo@runaway.xs4all.nl>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

#include "sys.h"
#include <stdio.h>
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif
#include <readline/tilde.h>  /* Part of libreadline that comes with binutils */
#include "bash.h"

static void print_usage(const char *progname)
{
  fprintf(stderr, "Usage: %s [options] [--] programname [...]\n", progname);
  fprintf(stderr, "Options: --version    Print version and exit successfully.\n");
  fprintf(stderr, "         --skip-dot   Skip directories in PATH that start with a dot.\n");
  fprintf(stderr, "         --skip-tilde Skip directories in PATH that start with a tilde.\n");
  fprintf(stderr, "         --show-dot   Don't expand a dot to current directory in output.\n");
  fprintf(stderr, "         --show-tilde Output a tilde for HOME directory for non-root.\n");
  fprintf(stderr, "         --tty-only   Stop processing options on the right if not on tty.\n");
  fprintf(stderr, "         --all, -a    Print all matching executables in PATH, not just the first\n");
}

static void print_version(void)
{
  fprintf(stdout, "GNU which v" VERSION ", Copyright (C) 1999 Carlo Wood.\n");
  fprintf(stdout, "GNU which comes with ABSOLUTELY NO WARRANTY;\n");
  fprintf(stdout, "This program is free software; your freedom to use, change\n");
  fprintf(stdout, "and distribute this program is covered by the GPL.\n");
}

static void print_fail(const char *progname, const char *name, const char *path_list)
{
  fprintf(stderr, "%s: no %s in (%s)\n", progname, name, path_list);
}

static int absolute_path_given;
static int found_path_starts_with_dot;
static char *abs_path;

static int skip_dot = 0, skip_tilde = 0, show_all = 0;

static char *find_command_in_path(const char *name, const char *path_list, int *path_index)
{
  char *found = NULL, *full_path;
  int status, name_len;

  name_len = strlen(name);

  if (!absolute_program(name))
    absolute_path_given = 0;
  else
  {
    char *p;
    absolute_path_given = 1;

    if (abs_path)
      free(abs_path);

    if (*name != '.' && *name != '/' && *name != '~')
    {
      abs_path = (char *)xmalloc(3 + name_len);
      strcpy(abs_path, "./");
      strcat(abs_path, name);
    }
    else
    {
      abs_path = (char *)xmalloc(1 + name_len);
      strcpy(abs_path, name);
    }

    path_list = abs_path;
    p = strrchr(abs_path, '/');
    *p++ = 0;
    name = p;
  }

  while (path_list && path_list[*path_index])
  {
    char *path;

    if (absolute_path_given)
    {
      path = savestring(path_list);
      *path_index = strlen(path);
    }
    else
      path = get_next_path_element(path_list, path_index);

    if (!path)
      break;

    if (*path == '~')
    {
      char *t = tilde_expand(path);
      free(path);
      path = t;

      if (skip_tilde)
      {
        free(path);
	continue;
      }
    }

    if (skip_dot && *path != '/')
    {
      free(path);
      continue;
    }

    found_path_starts_with_dot = (*path == '.');

    full_path = make_full_pathname(path, name, name_len);
    free(path);

    status = file_status(full_path);

    if ((status & FS_EXISTS) && (status & FS_EXECABLE))
    {
      found = full_path;
      break;
    }

    free(full_path);
  }

  return (found);
}

static char cwd[256];
static size_t cwdlen;

static void get_current_working_directory(void)
{
  if (cwdlen)
    return;

  if (!getcwd(cwd, sizeof(cwd)))
  {
    const char *pwd = getenv("PWD");
    if (pwd && strlen(pwd) < sizeof(cwd))
      strcpy(cwd, pwd);
  }

  if (*cwd != '/')
  {
    fprintf(stderr, "Can't get current working directory\n");
    exit(-1);
  }

  cwdlen = strlen(cwd);

  if (cwd[cwdlen - 1] != '/')
  {
    cwd[cwdlen++] = '/';
    cwd[cwdlen] = 0;
  }
}

static char *path_clean_up(const char *path)
{
  static char result[256];

  const char *p1 = path;
  char *p2 = result;

  int saw_slash = 0, saw_slash_dot = 0, saw_slash_dot_dot = 0;

  if (*p1 != '/')
  {
    get_current_working_directory();
    strcpy(result, cwd);
    saw_slash = 1;
    p2 = &result[cwdlen];
  }

  do
  {
    if (!saw_slash || *p1 != '/')
      *p2++ = *p1;
    if (saw_slash_dot && (*p1 == '/'))
      p2 -= 2;
    if (saw_slash_dot_dot && (*p1 == '/'))
    {
      int cnt = 0;
      do
      {
	if (--p2 < result)
	{
	  strcpy(result, path);
	  return result;
	}
	if (*p2 == '/')
	  ++cnt;
      }
      while (cnt != 3);
      ++p2;
    }
    saw_slash_dot_dot = saw_slash_dot && (*p1 == '.');
    saw_slash_dot = saw_slash && (*p1 == '.');
    saw_slash = (*p1 == '/');
  }
  while (*p1++);

  return result;
}

enum opts {
  opt_version,
  opt_skip_dot,
  opt_skip_tilde,
  opt_show_dot,
  opt_show_tilde,
  opt_tty_only
};

int main(int argc, char *argv[])
{
  const char *progname = argv[0];
  const char *path_list = getenv("PATH");
  char home[256];
  size_t homelen = 0;
  int short_option, long_option, fail_count = 0, path_index;
  int show_dot = 0, show_tilde = 0, tty_only = 0;
  struct option longopts[] = {
    {"version", 0, &long_option, opt_version},
    {"skip-dot", 0, &long_option, opt_skip_dot},
    {"skip-tilde", 0, &long_option, opt_skip_tilde},
    {"show-dot", 0, &long_option, opt_show_dot},
    {"show-tilde", 0, &long_option, opt_show_tilde},
    {"tty-only", 0, &long_option, opt_tty_only},
    {"all", 0, NULL, 'a'},
    {NULL, 0, NULL, 0}
  };

  while ((short_option = getopt_long(argc, argv, "a", longopts, NULL)) != -1)
  {
    switch (short_option)
    {
      case 0:
	switch (long_option)
	{
	  case opt_version:
	    print_version();
	    return 0;
	  case opt_skip_dot:
	    skip_dot = !tty_only;
	    break;
	  case opt_skip_tilde:
	    skip_tilde = !tty_only;
	    break;
	  case opt_show_dot:
	    show_dot = !tty_only;
	    break;
	  case opt_show_tilde:
	    show_tilde = (!tty_only && geteuid() != 0);
	    break;
	  case opt_tty_only:
	    tty_only = !isatty(1);
	    break;
	}
	break;
      case 'a':
        show_all = 1;
        break;
    }
  }

  if (show_dot)
    get_current_working_directory();

  if (show_tilde)
  {
    const char *h;

    if (!(h = getenv("HOME")))
    {
      fprintf(stderr, "%s: --show-tilde: Environment variable HOME not set\n", progname);
      show_tilde = 0;
    }
    else
    {
      strncpy(home, h, sizeof(home));
      home[sizeof(home) - 1] = 0;
      homelen = strlen(home);
      if (home[homelen - 1] != '/' && homelen < sizeof(home) - 1)
      {
	strcat(home, "/");
	++homelen;
      }
    }
  }

  argv += optind;

  if (!*argv)
  {
    print_usage(progname);
    return -1;
  }

  for (; *argv; ++argv)
  {
    char *result = NULL;
    int found_something = 0;

    if (path_list && *path_list != '\0')
    {
      path_index = 0;
      do
      {
	result = find_command_in_path(*argv, path_list, &path_index);
	if (result)
	{
	  const char *full_path = path_clean_up(result);
	  if (show_dot && found_path_starts_with_dot && !strncmp(full_path, cwd, cwdlen))
	  {
	    full_path += cwdlen;
	    fprintf(stdout, "./");
	  }
	  else if (show_tilde && !strncmp(full_path, home, homelen))
	  {
	    full_path += homelen;
	    fprintf(stdout, "~/");
	  }
	  fprintf(stdout, "%s\n", full_path);
	  free(result);
	  found_something = 1;
	}
      }
      while(show_all && result);
    }
    if (!found_something)
    {
      print_fail(progname, absolute_path_given ? strrchr(*argv, '/') + 1 : *argv, absolute_path_given ? abs_path : path_list);
      ++fail_count;
    }
  }

  return fail_count;
}
